### PR TITLE
[IMP] web_editor, *: make media-dialog search odoo media-library

### PR DESCRIPTION
**Pull request by sdegueldre**
*: web_unsplash

We have recently created an odoo media library that exposes a public API
for searching for open-source media, including dynamic SVGs that can be
used with the dynamic background-shape system to create illustrations
whose colors can be changed by the user. This commit adds the
corresponding client-side functionalty: search for the images in the
media-library, creating a corresponding attachment when selected, as
well as an option to customize the colors of the dynamic SVGs.

In the process, adaptations have been made in the web_unsplash module to
avoid duplicating code and to adapt code that was making assumptions
that are no longer valid.

task-2316651
Linked to: odoo/iap-apps#235

### CODE REVIEW

File: addons/web_editor/static/src/js/editor/snippets.options.js
Comment (JKE-be) [inline addons/web_editor/static/src/js/editor/snippets.options.js:3584]: quid about cdn ? lazyloading? ...

File: addons/web_editor/controllers/main.py
Comment (JKE-be) [inline addons/web_editor/controllers/main.py:586]: And why not use our classic slug if you want to concat name and id...

File: addons/web_editor/controllers/main.py
Comment (JKE-be) [inline addons/web_editor/controllers/main.py:560]: @mart-e 
What's the last security rule for SVG ?

File: addons/web_editor/static/src/js/editor/snippets.options.js
Comment (sdegueldre) [reply addons/web_editor/static/src/js/editor/snippets.options.js:3584]: Not sure what you mean here. Dynamic SVGs are always local.

File: addons/web_editor/controllers/main.py
Comment (sdegueldre) [reply addons/web_editor/controllers/main.py:586]: I need the url to be `/web_editor/shape/...` to hit the color-customizing controller

File: addons/web_editor/static/src/js/editor/snippets.options.js
Comment (JKE-be) [reply addons/web_editor/static/src/js/editor/snippets.options.js:3584]: pq ca ne pourrait pas utiliser le cdn ?

File: addons/web_unsplash/static/src/js/unsplash_image_widget.js
Comment (JKE-be) [inline addons/web_unsplash/static/src/js/unsplash_image_widget.js:123]: Could we rename it to hasContent instead of negative function name

File: addons/web_editor/controllers/main.py
Comment (JKE-be) [reply addons/web_editor/controllers/main.py:560]: > Olivier (odo): - si c'est stocké dans un ir.attachment, lors du create() on verifie que user._is_system() et on force le mimetype en text/plain (ce qui en theorie casserait les SVG)
> 
> Olivier (odo): - si c'est stocké dans un fields qui n'est pas en attachment, alors on check record.env.is_system() et si c'est pas le cas on raise une error
> 

=> donc ca ne doit pas marché quand t'es pas admin dans l'etat.

File: addons/web_editor/controllers/main.py
Comment (JKE-be) [reply addons/web_editor/controllers/main.py:586]: external_shapes is a module that's a depends of web_editor ? not sure

no I mean: `name-id`  instead of `id_name` (standard slug in Odoo)

Comment (sdegueldre) [reply addons/web_editor/controllers/main.py:586]: external_shapes in not a module, this leverages the attachment fallback from the `/web_editor/shapes` route

Changed to name-id, didn't know we had a convention.

File: addons/web_editor/static/src/js/editor/snippets.options.js
Comment (sdegueldre) [reply addons/web_editor/static/src/js/editor/snippets.options.js:3584]: We need to replace the colors inside the SVG. If we do end up using a CDN that also replaces the colors in the SVG then this option would need to take that into account in its computeWidgetVisibility method but I don't see why this line would be a problem.

File: addons/web_editor/controllers/main.py
Comment (sdegueldre) [reply addons/web_editor/controllers/main.py:560]: Indeed, just tried to edit as a designer and it failed, attachment did come back as `text/plain`. Should I check the urls against a white-list of trusted origins and create the attachments with sudo?

Comment (JKE-be) [reply addons/web_editor/controllers/main.py:560]: we could imagine to bypass it for website_designer but quid in all others cases where end user will try...

Probably that just whitelist our media library and make sudo is a possibility
+ request of odony is to avoid call from browser to IAP in this case, but make it s2s directly and avoid to leak the dbuuid.

File: addons/web_editor/controllers/main.py
Comment (JKE-be) [inline addons/web_editor/controllers/main.py:555]: my bad, forgot to specify, will be better if it can be done in post, to avoid dbuid in log too

File: addons/web_editor/controllers/main.py
Comment (JKE-be) [inline addons/web_editor/controllers/main.py:576]: ','.join(media.keys())

File: addons/web_editor/controllers/main.py
Comment (JKE-be) [inline addons/web_editor/controllers/main.py:584]: raise exception, will pobably already log something gin the log.
And avoid duplicate string, you can just use var if my first supposition is not true

File: addons/web_editor/controllers/main.py
Comment (JKE-be) [inline addons/web_editor/controllers/main.py:589]: I love the latin, but since the medium have another meaning in marketing terms, can se simple use an other name ;)

btw since only tje download_urls is whitelisted, we should only create attachment for returned items.

So:
```python
for id, url in download_urls:
    req = requests.get(url)
    name = '_'.join([media[id]['query'], url.split('/')[-1]])
    attachment = request.env['ir.attachment'].create({ ... })
    if media[id].get('is_dynamic_svg''):
        attachment['url'] = '/web_editor/shape/external_shapes/%s' % slug(attachment)
```

File: addons/web_editor/controllers/main.py
Comment (sdegueldre) [reply addons/web_editor/controllers/main.py:584]: I took the same approach as in web_unsplash, I thought the idea was that the first line wouldn't be translated, but since the line raising the exception is logged in the traceback, there isn't much of a point. Removed the duplication :)

File: addons/website_blog/static/src/js/tours/website_blog.js
Comment (sdegueldre) [reply addons/website_blog/static/src/js/tours/website_blog.js:35]: No reliance on the service during testing: we don't query the library if the search field is empty. I have to trigger an input event on the search field because of FP's request that we don't show default images when you open the media-dialog, but when searching for nothing it will only query the database and not external services.

**Comment by JKE-be** [comment]
quid about caching ?  unique should be keept for cache, c1 will btw alrady invalidate if it is changed.

image

File: addons/web_editor/static/tests/field_html_tests.js
Comment (sdegueldre) [reply addons/web_editor/static/tests/field_html_tests.js:335]: I just removed this because we changed the library querying to be s2s, but before this it had to be a valid URL (doesn't matter if it's unreachable) but this is no longer relevant as I removed it.

File: addons/web_editor/static/src/xml/wysiwyg.xml
Comment (sdegueldre) [reply addons/web_editor/static/src/xml/wysiwyg.xml:129]: Probably messed something up when I refactored noContent -> hasContent, I'll look into it.

**Comment by sdegueldre** [comment]
> quid about caching ? unique should be keept for cache, c1 will btw alrady invalidate if it is changed.
> 
> image

Not sure what you mean here. The data-original-src is caused by the same bug that lets the width/quality options be shown on SVGs, it simply shouldn't be there at all, I still need to fix this in stable :/

**Comment by JKE-be** [comment]
> > quid about caching ? unique should be keept for cache, c1 will btw alrady invalidate if it is changed.
> > image
> 
> Not sure what you mean here. The data-original-src is caused by the same bug that lets the width/quality options be shown on SVGs, it simply shouldn't be there at all, I still need to fix this in stable :/

the end user should have the uniq id in the url, to benefit of the cache browser.

**Comment by sdegueldre** [comment]
> > > quid about caching ? unique should be keept for cache, c1 will btw alrady invalidate if it is changed.
> > > image
> > 
> > 
> > Not sure what you mean here. The data-original-src is caused by the same bug that lets the width/quality options be shown on SVGs, it simply shouldn't be there at all, I still need to fix this in stable :/
> 
> the end user should have the uniq id in the url, to benefit of the cache browser.

The `/web_editor/shapes` controller responses have their max-age header set to http.STATIC_CACHE_LONG, they don't go through the normal attachment resolution path.

File: addons/web_editor/static/src/js/wysiwyg/widgets/media.js
Comment (JKE-be) [inline addons/web_editor/static/src/js/wysiwyg/widgets/media.js:309]: imo /%/static/% if we follow idea of fp ! no image from any module :/

File: addons/web_editor/static/src/js/wysiwyg/widgets/media.js
Comment (JKE-be) [inline addons/web_editor/static/src/js/wysiwyg/widgets/media.js:752]: can you double check that really translatable please and confirm. thx

**Review by JKE-be — COMMENTED** [review]
what's about our yesterday conversation an a centralized controller to search in one rcp instead of 3 ?

File: addons/web_editor/controllers/main.py
Comment (odony) [inline addons/web_editor/controllers/main.py:515]: unrelated, but please never use `open()` directly -> `tools.file_open()`

File: addons/web_editor/static/src/js/wysiwyg/widgets/media.js
Comment (sdegueldre) [reply addons/web_editor/static/src/js/wysiwyg/widgets/media.js:752]: Can confirm this does not work: babel treats the entire content of the string literal as a string, even the parts between interpolation sigils. I've extracted the translation outside the template literal and that fixes it, but this will probably get removed anyway and go directly in the XML template once I figure out how to not show any of the default images.

File: addons/web_editor/controllers/main.py
Comment (sdegueldre) [reply addons/web_editor/controllers/main.py:600]: Then I'm not sure what you mean. Are you saying I should be using `ir_http slugify`?

File: addons/website/views/snippets/s_image_text.xml
Comment (JKE-be) [inline addons/website/views/snippets/s_image_text.xml:9]: as explain, it will break all overridden of theme ! 
Each theme replace the xml id s_image_text_default_image with there own image:

```xml
    <record id="s_text_image_default_image" model="theme.ir.attachment">
        <field name="key">website.s_text_image_default_image</field>
        <field name="name">website.s_text_image_default_image</field>
        <field name="url">/theme_artists/static/src/img/snippets/s_text_image.jpg</field>
    </record>
    <record id="s_image_text_default_image" model="theme.ir.attachment">
        <field name="key">website.s_image_text_default_image</field>
        <field name="name">website.s_image_text_default_image</field>
        <field name="url">/theme_artists/static/src/img/snippets/s_image_text.jpg</field>
    </record>
```

**Review by JKE-be — CHANGES_REQUESTED** [review]
Some point to check, fix or get more insights

+ please fix tour as discuss Wednesday... Else we cannot merge it...

+ Waiting approval of Odony about the hack to allow end user to upload svg.
Reminder: only from our own endpoint (iap services) where we control the content.

**Comment by JKE-be** [comment]
+ can we really be more strict with input c[1-6], Imo we shoudl just support #[/d]{6|8}
here we inject input whatever the len, with just an html encode.
I have no doubt that in a few months we will find a new pattern to inject similiar to url(base64,svg+xml;bsbsbsbs=)  that will execute b64 js code

File: addons/web_editor/static/tests/test_utils.js
Comment (sdegueldre) [reply addons/web_editor/static/tests/test_utils.js:278]: Because of the new /static/ filter in the media-dialog, this image would no longer show up with the /web/static/ url

**Comment by sdegueldre** [comment]
@JKE-be I would like to support rgba at least since that's what the colorpickers output when the color is partially transparent, but I can make it a stringent check for a hex or rgb(a) color if that's ok with you.

File: addons/web_editor/static/src/xml/wysiwyg.xml
Comment (sdegueldre) [reply addons/web_editor/static/src/xml/wysiwyg.xml:325]: Adding the parameter with string interpolation was causing invalid URLs if the image_src already had params, so I moved the insertion of this param to media.js line 782, but indeed the comment should move there too.

File: addons/web_editor/controllers/main.py
Comment (sdegueldre) [reply addons/web_editor/controllers/main.py:517]: Public user cannot search attachments without sudo, even with public = True in the domain. Is this not the expected behavior?

File: addons/website/views/snippets/s_image_text.xml
Comment (sdegueldre) [reply addons/website/views/snippets/s_image_text.xml:9]: As I explained when we talked about it previously, since the illustrations need a specific url for their colors to be customizable, it cannot use the current xml_id customization system, and I currently don't have any idea how to make them both dynamic and customizable by the theme.

Comment (JKE-be) [reply addons/website/views/snippets/s_image_text.xml:9]: yes, but xml id it a redirection to an url. It is the xml id website.s_image_text_default_image that you need to update and add your custom url in field url.

```python
curl -D -  https://www.odoo.com/web/image/website.s_image_text_default_image
HTTP/1.1 301 MOVED PERMANENTLY
server: nginx
date: Mon, 14 Sep 2020 08:16:46 GMT
content-type: text/html; charset=utf-8
content-length: 315
location: https://www.odoo.com/website/static/src/img/snippets_demo/s_image_text.png
```

File: addons/web_editor/controllers/main.py
Comment (JKE-be) [reply addons/web_editor/controllers/main.py:517]: yes it is the case since this commit: 62c9dedafda0cbdc618641eaa80c49535ab0b5f6

Seems legit with the is_public = True, but not sure how your previous code was working in public user in this case.

File: addons/website/views/snippets/s_image_text.xml
Comment (sdegueldre) [reply addons/website/views/snippets/s_image_text.xml:9]: I understand that, but it still makes the url in the page non-customizable from the editor. That means that the user cannot change the colors, and I cannot use the main color when dropping the snippet in the page. While the url pointed to by the redirect could be a color-customized shape url, the editor would have no access to it.

Comment (JKE-be) [reply addons/website/views/snippets/s_image_text.xml:9]: ok, now I catch your problem...  
Probably that the right tradeoff solution is to don't have default svg in website, and that theme replace with xpath the image wth xpath instead to use the url mechanism if svg is a 'requirement' for theme (e.g. paptic)

**Comment by JKE-be** [comment]
yes no problem to support one or other case, but I mean restrict to what we want to support.
Just don't keep the door open to any strange way to do the same even if it isvalid. We just need to support what we are expected to receive as arguments. thx

File: addons/website/views/snippets/s_image_text.xml
Comment (sdegueldre) [reply addons/website/views/snippets/s_image_text.xml]: OK. Removed illustrations from website snippets, themes can add illustrations if they need them with an xpath.

File: odoo/modules/module.py
Comment (JKE-be) [inline odoo/modules/module.py:187]: leftover

File: addons/website_blog/tests/test_ui.py
Comment (JKE-be) [inline addons/website_blog/tests/test_ui.py:15]: can we have a real picture, to avoid potential 404 error and more in log ?

Comment (sdegueldre) [reply addons/website_blog/tests/test_ui.py:15]: does this not simply return the placeholder image when there is no attachment corresponding to the id? I don't get any 404 in the logs with this, regardless of the id. If you still want me to use a real image, do you have a suggestion that isn't a /static/ image? Since it needs to show up in the media-dialog

File: addons/web_editor/static/src/js/wysiwyg/widgets/media.js
Comment (JKE-be) [inline addons/web_editor/static/src/js/wysiwyg/widgets/media.js:150]: @sdegueldre 

IS there any good reason to don't use the same type of structure everywhere ? 
Seems like is complexifed due to this list vs set.

Eg: https://github.com/odoo/odoo/pull/56479/files#diff-194923481b66ad4811cdca3dc7b96693R480

It will be probably more clean if we use the same tructure. What do you think about ?

File: addons/web_editor/static/src/js/wysiwyg/widgets/media.js
Comment (JKE-be) [inline addons/web_editor/static/src/js/wysiwyg/widgets/media.js:903]: This default color of undraw should be probably a constant instead to have it duplicated.
Doesn't make it sense to replace it by enterprise color on the iap service directly - to be more Odoo branded

Comment (JKE-be) [reply addons/web_editor/static/src/js/wysiwyg/widgets/media.js:903]: And in case we add a second library, we will need to choose a default color for substitution... so probably than it is now the good timing to choose it.

Comment (JKE-be) [reply addons/web_editor/static/src/js/wysiwyg/widgets/media.js:903]: #00A09D if we want to keep same green style, else: #875A7B ?

Comment (sdegueldre) [reply addons/web_editor/static/src/js/wysiwyg/widgets/media.js:903]: This color is always replaced when coming from IAP, it's never shown as-is so you would not see a difference even if it was the enterprise color. It's also the color that's replaced by the main theme color in the shape controller, so I would need to pre-process the SVGs when downloading them from IAP to make them compatible. Overall there is no benefit.

I will extract it to a constant though.

File: addons/web_editor/static/src/js/wysiwyg/widgets/media.js
Comment (sdegueldre) [reply addons/web_editor/static/src/js/wysiwyg/widgets/media.js:150]: Seems your link isn't working. I guess you're wondering why I use a set for selectedMedia and an array for libraryMedia? It's because I need to iterate on libraryMedia to render them, and it's what's returned directly by the API, but for selected media I need to check whether it's already selected or not and toggle it which is much cleaner with a set (rather than splicing an array)

File: addons/web_editor/static/src/js/wysiwyg/widgets/media.js
Comment (JKE-be) [reply addons/web_editor/static/src/js/wysiwyg/widgets/media.js:903]: imo it is not during download, but during import on iap server that we need to choose a color.

whatever the color, we need a standard for the next one library too.

So if we want to keep arbitrary #3AADAA why not. Just another new 'constant' to keep in mind ;)

File: addons/web_editor/static/src/js/wysiwyg/widgets/media.js
Comment (JKE-be) [reply addons/web_editor/static/src/js/wysiwyg/widgets/media.js:150]: selectedAttachments / selectedMedia

here the kind og code that should be simplified using the same structure...

image


Btw, a set is iterable like an list.
Maybe @qsm-odoo has another opinion than me ?

File: addons/web_editor/static/src/js/wysiwyg/widgets/media.js
Comment (sdegueldre) [reply addons/web_editor/static/src/js/wysiwyg/widgets/media.js:903]: The default colors for substitution were already chosen during the shapes task, and we went with the colors of the default palette so that designers could design their shapes with the default palette and know how they would look like.

Comment (JKE-be) [reply addons/web_editor/static/src/js/wysiwyg/widgets/media.js:903]: ok in this case, lets go to keep it ;) jsute make it as constant. thx

File: addons/web_editor/static/src/js/wysiwyg/widgets/media.js
Comment (sdegueldre) [reply addons/web_editor/static/src/js/wysiwyg/widgets/media.js:150]: Right I see what you're saying now. Considering the number of selected media will never be huge in practice I guess I can just go with an array and keep most of the existing code untouched.

File: addons/web_editor/controllers/main.py
Comment (JKE-be) [inline addons/web_editor/controllers/main.py:570]: is it really optional media ? 
because it will crash a the first use with media.keys()

File: addons/web_editor/static/src/js/wysiwyg/widgets/media.js
Comment (JKE-be) [inline addons/web_editor/static/src/js/wysiwyg/widgets/media.js:749]: I reported a bug on discord a few day ago when you click on some static image where image_src was null, it has been processed and fixed ?

Comment (sdegueldre) [reply addons/web_editor/static/src/js/wysiwyg/widgets/media.js:749]: Yes, that was caused by attachments being created with the wrong mimetype when not logged in as admin, they were not considered images and hence had no image_src, but this has been fixed.

**Comment by JKE-be** [comment]
@robodoo r+

**Comment by JKE-be** [comment]
@robodoo retry

**Comment by JKE-be** [comment]
@robodoo r+

**Pull request closed without merging**
